File: easycontrol/server/src/main/java/top/saymzx/easycontrol/server/wrappers/SurfaceControl.java
/*
 * 本项目大量借鉴学习了开源投屏软件：Scrcpy，在此对该项目表示感谢
 */
package top.saymzx.easycontrol.server.wrappers;

import android.annotation.SuppressLint;
import android.graphics.Rect;
import android.os.Build;
import android.os.IBinder;
import android.view.Surface;

import java.lang.reflect.InvocationTargetException;
import java.lang.reflect.Method;

@SuppressLint("PrivateApi")
public final class SurfaceControl {

  private static Class<?> CLASS;

  private static Method getBuiltInDisplayMethod = null;
  private static Method setDisplayPowerModeMethod = null;
  private static Method getPhysicalDisplayTokenMethod = null;
  private static Method getPhysicalDisplayIdsMethod = null;

  public static void init() throws ClassNotFoundException {
    CLASS = Class.forName("android.view.SurfaceControl");
    try {
      if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.Q) {
        try {
          getPhysicalDisplayIdsMethod = CLASS.getMethod("getPhysicalDisplayIds");
          getPhysicalDisplayTokenMethod = CLASS.getMethod("getPhysicalDisplayToken", long.class);
        } catch (Exception ignored) {
          getMethodAndroid14();
        }
      }
      setDisplayPowerModeMethod = CLASS.getMethod("setDisplayPowerMode", IBinder.class, int.class);
    } catch (Exception ignored) {
    }
  }

  // 安卓14之后部分函数转移到了DisplayControl
  @SuppressLint({"PrivateApi", "SoonBlockedPrivateApi", "BlockedPrivateApi"})
  private static void getMethodAndroid14() throws Exception {
    Class<?> displayControlClass = null;
    try {
      Method createClassLoaderMethod = Class.forName("com.android.internal.os.ClassLoaderFactory").getDeclaredMethod("createClassLoader", String.class, String.class, String.class, ClassLoader.class, int.class, boolean.class, String.class);
      ClassLoader classLoader = (ClassLoader) createClassLoaderMethod.invoke(null, "/system/framework/services.jar", null, null, ClassLoader.getSystemClassLoader(), 0, true, null);
      displayControlClass = classLoader.loadClass("com.android.server.display.DisplayControl");
//      Method loadMethod = Runtime.class.getDeclaredMethod("loadLibrary0", Class.class, String.class);
//      loadMethod.setAccessible(true);
//      if ((Build.BRAND.toLowerCase() + Build.MANUFACTURER.toLowerCase()).contains("honor")) throw new Exception("Honor device");
//      loadMethod.invoke(Runtime.getRuntime(), displayControlClass, "android_servers");
    } catch (Throwable ignored) {
    }
    if (displayControlClass == null) throw new Exception("Failed to load DisplayControl class");
    getPhysicalDisplayIdsMethod = displayControlClass.getMethod("getPhysicalDisplayIds");
    getPhysicalDisplayTokenMethod = displayControlClass.getMethod("getPhysicalDisplayToken", long.class);
  }

  public static void openTransaction() throws NoSuchMethodException, InvocationTargetException, IllegalAccessException {
    CLASS.getMethod("openTransaction").invoke(null);
  }

  public static void closeTransaction() throws NoSuchMethodException, InvocationTargetException, IllegalAccessException {
    CLASS.getMethod("closeTransaction").invoke(null);
  }

  public static void setDisplayProjection(IBinder displayToken, int orientation, Rect layerStackRect, Rect displayRect) throws NoSuchMethodException, InvocationTargetException, IllegalAccessException {
    CLASS.getMethod("setDisplayProjection", IBinder.class, int.class, Rect.class, Rect.class).invoke(null, displayToken, orientation, layerStackRect, displayRect);
  }

  public static void setDisplayLayerStack(IBinder displayToken, int layerStack) throws NoSuchMethodException, InvocationTargetException, IllegalAccessException {
    CLASS.getMethod("setDisplayLayerStack", IBinder.class, int.class).invoke(null, displayToken, layerStack);
  }

  public static void setDisplaySurface(IBinder displayToken, Surface surface) throws NoSuchMethodException, InvocationTargetException, IllegalAccessException {
    CLASS.getMethod("setDisplaySurface", IBinder.class, Surface.class).invoke(null, displayToken, surface);
  }

  public static IBinder createDisplay(String name, boolean secure) throws NoSuchMethodException, InvocationTargetException, IllegalAccessException {
    return (IBinder) CLASS.getMethod("createDisplay", String.class, boolean.class).invoke(null, name, secure);
  }

  public static void destroyDisplay(IBinder displayToken) throws NoSuchMethodException, InvocationTargetException, IllegalAccessException {
    CLASS.getMethod("destroyDisplay", IBinder.class).invoke(null, displayToken);
  }

  public static IBinder getBuiltInDisplay() {
    try {
      if (getBuiltInDisplayMethod == null) getBuiltInDisplayMethod = CLASS.getMethod("getBuiltInDisplay", int.class);
      return (IBinder) getBuiltInDisplayMethod.invoke(null, 0);
    } catch (Exception ignored) {
      return null;
    }
  }

  public static IBinder getPhysicalDisplayToken(long physicalDisplayId) {
    if (getPhysicalDisplayTokenMethod == null) return null;
    try {
      return (IBinder) getPhysicalDisplayTokenMethod.invoke(null, physicalDisplayId);
    } catch (Exception ignored) {
      return null;
    }
  }

  public static long[] getPhysicalDisplayIds() {
    if (getPhysicalDisplayIdsMethod == null) return null;
    try {
      return (long[]) getPhysicalDisplayIdsMethod.invoke(null);
    } catch (Exception ignored) {
      return null;
    }
  }

  public static void setDisplayPowerMode(IBinder displayToken, int mode) {
    if (setDisplayPowerModeMethod == null) return;
    try {
      setDisplayPowerModeMethod.invoke(null, displayToken, mode);
    } catch (Exception ignored) {
    }
  }

}

File: scrcpy_android/server/src/main/java/top/saymzx/scrcpy/server/wrappers/SurfaceControl.kt
package top.saymzx.scrcpy.server.wrappers

import android.annotation.SuppressLint
import android.graphics.Rect
import android.os.Build
import android.os.IBinder
import android.view.Surface
import java.lang.reflect.InvocationTargetException

@SuppressLint("PrivateApi")
object SurfaceControl {
  private var CLASS = Class.forName("android.view.SurfaceControl")

  const val POWER_MODE_OFF = 0
  const val POWER_MODE_NORMAL = 2

  private var getBuiltInDisplayMethod = if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
    CLASS.getMethod("getBuiltInDisplay", Int::class.javaPrimitiveType)
  } else {
    CLASS.getMethod("getInternalDisplayToken")
  }

  private var setDisplayPowerModeMethod = CLASS.getMethod(
    "setDisplayPowerMode",
    IBinder::class.java,
    Int::class.javaPrimitiveType
  )

  private var getPhysicalDisplayTokenMethod =
    CLASS.getMethod("getPhysicalDisplayToken", Long::class.javaPrimitiveType)

  private var getPhysicalDisplayIdsMethod = CLASS.getMethod("getPhysicalDisplayIds")

  fun openTransaction() {
    try {
      CLASS.getMethod("openTransaction").invoke(null)
    } catch (e: Exception) {
      throw AssertionError(e)
    }
  }

  fun closeTransaction() {
    try {
      CLASS.getMethod("closeTransaction").invoke(null)
    } catch (e: Exception) {
      throw AssertionError(e)
    }
  }

  fun setDisplayProjection(
    displayToken: IBinder?,
    orientation: Int,
    layerStackRect: Rect?,
    displayRect: Rect?
  ) {
    try {
      CLASS.getMethod(
        "setDisplayProjection",
        IBinder::class.java,
        Int::class.javaPrimitiveType,
        Rect::class.java,
        Rect::class.java
      )
        .invoke(null, displayToken, orientation, layerStackRect, displayRect)
    } catch (e: Exception) {
      throw AssertionError(e)
    }
  }

  fun setDisplayLayerStack(displayToken: IBinder?, layerStack: Int) {
    try {
      CLASS.getMethod("setDisplayLayerStack", IBinder::class.java, Int::class.javaPrimitiveType)
        .invoke(null, displayToken, layerStack)
    } catch (e: Exception) {
      throw AssertionError(e)
    }
  }

  fun setDisplaySurface(displayToken: IBinder?, surface: Surface?) {
    try {
      CLASS.getMethod("setDisplaySurface", IBinder::class.java, Surface::class.java)
        .invoke(null, displayToken, surface)
    } catch (e: Exception) {
      throw AssertionError(e)
    }
  }

  fun createDisplay(name: String?, secure: Boolean): IBinder {
    return try {
      CLASS.getMethod("createDisplay", String::class.java, Boolean::class.javaPrimitiveType)
        .invoke(null, name, secure) as IBinder
    } catch (e: Exception) {
      throw AssertionError(e)
    }
  }
  // call getBuiltInDisplay(0)

  // call getInternalDisplayToken()
  val builtInDisplay: IBinder?
    get() = try {
      val method = getBuiltInDisplayMethod
      if (Build.VERSION.SDK_INT < Build.VERSION_CODES.Q) {
        // call getBuiltInDisplay(0)
        method.invoke(null, 0) as IBinder
      } else method.invoke(null) as IBinder

      // call getInternalDisplayToken()
    } catch (e: InvocationTargetException) {
      null
    } catch (e: IllegalAccessException) {
      null
    } catch (e: NoSuchMethodException) {
      null
    }

  fun getPhysicalDisplayToken(physicalDisplayId: Long): IBinder? {
    return try {
      val method = getPhysicalDisplayTokenMethod
      method.invoke(null, physicalDisplayId) as IBinder
    } catch (e: InvocationTargetException) {
      null
    } catch (e: IllegalAccessException) {
      null
    } catch (e: NoSuchMethodException) {
      null
    }
  }

  val physicalDisplayIds: LongArray?
    get() = try {
      val method = getPhysicalDisplayIdsMethod
      method.invoke(null) as LongArray
    } catch (e: InvocationTargetException) {
      null
    } catch (e: IllegalAccessException) {
      null
    } catch (e: NoSuchMethodException) {
      null
    }

  fun setDisplayPowerMode(displayToken: IBinder?, mode: Int): Boolean {
    return try {
      val method = setDisplayPowerModeMethod
      method.invoke(null, displayToken, mode)
      true
    } catch (e: InvocationTargetException) {
      false
    } catch (e: IllegalAccessException) {
      false
    } catch (e: NoSuchMethodException) {
      false
    }
  }

  fun destroyDisplay(displayToken: IBinder?) {
    try {
      CLASS.getMethod("destroyDisplay", IBinder::class.java).invoke(null, displayToken)
    } catch (e: Exception) {
      throw AssertionError(e)
    }
  }
}